backend/path-project.go

```diff
@@ -141,3 +156,41 @@ func handleProjectUpdate(ctx context.Context, req *logical.Request, data *framew
 		Data: item.Data(),
 	}, nil
 }
+
+func handleProjectDelete(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	projectName := data.Get("project").(string)
+	project, err := loadProject(ctx, req.Storage, projectName)
+	if err != nil {
+		return nil, err
+	}
+
+	if project == nil {
+		return logical.ErrorResponse("project %s is not configured in vault", projectName), nil
+	}
+
+	// Remove project name from vault
+	err = req.Storage.Delete(ctx, KeyProjectConfigPrefix+projectName)
+	if err != nil {
+		return nil, err
+	}
+
+	// Cleanup project dsn entries from vault too
+	keys, err := req.Storage.List(ctx, KeyDsnPrefix+projectName+"/")
+	if err != nil {
+		return nil, err
+	}
+
+	for _, key := range keys {
+		err = req.Storage.Delete(ctx, KeyDsnPrefix+projectName+"/"+key)
+		if err != nil {
+			return nil, err
```

Comment:
there is a bug here if this fails halfway and we return the error.

Vault will consider the delete operation a failure, but we've already deleted the project on line 172. There is no way to clean up the remaining DSNs afterwards. all subsequent attempts to delete the project will fail on line 168.

A better way to approach this is to keep going through the list and collect errors in a list. Then return a success response with warnings if we get any error.

The `*logical.Response` object has a method to add warnings that you can use.

Reply:
IMO, first we have to delete the child keys and after only successful deletion of all the child keys we can delete the main project key.

Reply:
yeah I think that would be a better way.

Reply:
Changes pushed